File: backend/Controllers/FriendshipController.cs
using backend.Data;
using backend.Models;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Identity;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using System.Security.Claims;
using System.Threading.Tasks;

namespace backend.Controllers
{
    [Authorize]
    [Route("api/[controller]")]
    [ApiController]
    public class FriendshipController : ControllerBase
    {
        private readonly BackendContext _context;
        private readonly UserManager<IdentityUser> _userManager;

        public FriendshipController(BackendContext context, UserManager<IdentityUser> userManager)
        {
            _context = context;
            _userManager = userManager;
        }

        // Send a friend request
        [HttpPost("send-request")]
        public async Task<IActionResult> SendFriendRequest([FromBody] FriendshipRequestModel model)
        {
            var user1 = await _userManager.FindByIdAsync(model.User1Id);
            var user2 = await _userManager.FindByIdAsync(model.User2Id);

            if (user1 == null || user2 == null)
            {
                return BadRequest("One or both users not found.");
            }

            var existingFriendship = await _context.Friendships
                .FirstOrDefaultAsync(f => (f.User1Id == model.User1Id && f.User2Id == model.User2Id) ||
                                          (f.User1Id == model.User2Id && f.User2Id == model.User1Id));

            if (existingFriendship != null)
            {
                return BadRequest("Friendship request already exists or the users are already friends.");
            }

            var friendship = new Friendship
            {
                User1Id = model.User1Id,
                User2Id = model.User2Id,
                IsAccepted = false // Friendship is pending
            };

            _context.Friendships.Add(friendship);
            await _context.SaveChangesAsync();

            return Ok(new { message = "Friend request sent successfully." });
        }

        // Accept a friend request
        [HttpPost("accept-request")]
        public async Task<IActionResult> AcceptFriendRequest([FromBody] FriendshipRequestModel model)
        {
            var friendship = await _context.Friendships
                .FirstOrDefaultAsync(f => f.User1Id == model.User1Id && f.User2Id == model.User2Id);

            if (friendship == null || friendship.IsAccepted)
            {
                return BadRequest("No pending friend request found or friendship already accepted.");
            }

            friendship.IsAccepted = true;

            _context.Friendships.Update(friendship);
            await _context.SaveChangesAsync();

            return Ok(new { message = "Friendship accepted." });
        }

        // Get the list of friends for a specific user
        [HttpGet("get-friends/{userId}")]
        public async Task<IActionResult> GetFriends(string userId)
        {
            var friendships = await _context.Friendships
                .Where(f => (f.User1Id == userId || f.User2Id == userId) && f.IsAccepted)
                .ToListAsync();

            var friends = friendships.Select(f =>
                f.User1Id == userId ? f.User2Id : f.User1Id).ToList();

            return Ok(friends);
        }

        // Check if two users are friends
        [HttpGet("are-friends/{userId1}/{userId2}")]
        public async Task<IActionResult> AreFriends(string userId1, string userId2)
        {
            var friendship = await _context.Friendships
                .FirstOrDefaultAsync(f => (f.User1Id == userId1 && f.User2Id == userId2) ||
                                          (f.User1Id == userId2 && f.User2Id == userId1));

            if (friendship != null && friendship.IsAccepted)
            {
                return Ok(new { message = "They are friends." });
            }

            return Ok(new { message = "They are not friends." });
        }
    }

    // Model for sending a friendship request
    public class FriendshipRequestModel
    {
        public string User1Id { get; set; }
        public string User2Id { get; set; }
    }
}

File: backend/controllers/friendshipController.js
import { sql, poolPromise } from '../db.js'; // adjust the path as needed

export const addFriend = async (req, res) => {

    const { username  } = req.body;

    const userId = req.user.id;

    if(!username){
        return res(400).json({ message: 'missing username!' });  
    }
    
    try{

        const pool = await poolPromise;
        const response = await pool
            .request()
            .input('Username', sql.NVarChar, username)
            .query(`
                Select Id FROM Users Where Username = @Username
            `);


        const friendId = response.recordset[0].Id;


        if(!friendId){
            return res.status(400).json({ message: 'Friend not found!' });  
        }
        

        await pool
            .request()
            .input('UserId', sql.Int, userId )
            .input('FriendId', sql.Int, friendId)
            .query(`
                INSERT INTO Friendships (UserId, FriendId)
                VALUES (@UserId, @FriendId)
            `);

        return res.status(200).json({ message: 'Friend request sent successfully' });
    }catch(err){

        console.error('add member error:', err);
        return res.status(500).json({ message: 'Server error' });
    }

}

export const getRequests = async (req, res) => {
    const id = req.user.id;

    console.log("getRequests route hit");

    try {
        const pool = await poolPromise;

        const result = await pool
            .request()
            .input('UserId', sql.Int, id)
            .query(`
                SELECT 
                    f.Id AS FriendshipId,
                    f.UserId AS SenderId,
                    u.Username AS SenderUsername,
                    f.Status,
                    f.CreatedAt
                FROM Friendships f
                JOIN Users u ON f.UserId = u.Id
                WHERE f.FriendId = @UserId AND f.Status = 'pending'
            `);

        return res.status(200).json({ requests: result.recordset });

    } catch (err) {
        console.error('Get requests error:', err);
        return res.status(500).json({ message: 'Server error' });
    }
};

export const acceptRequest = async (req, res) => {

    const { userId } = req.body;

    if(!userId){
        return res.status(400).json({ message: 'missing request id!' });  
    }

    const friendId = req.user.id;
    
    try {
        const pool = await poolPromise;

        const result = await pool
            .request()
            .input('UserId', sql.Int, userId)
            .input('FriendId', sql.Int, friendId)
            .query(`
                UPDATE Friendships
                SET Status = 'accepted'
                WHERE UserId = @UserId AND FriendId = @FriendId AND Status = 'pending'
            `);


        return res.status(200).json({ userId, message: 'Friend request accepted successfully' });

    } catch (err) {
        console.error('Get requests error:', err);
        return res.status(500).json({ message: 'Server error' });
    }
}

export const getFriends = async (req, res) => {
    const id = req.user.id;

    try {
        const pool = await poolPromise;

        const result = await pool
            .request()
            .input('UserId', sql.Int, id)
            .query(`
                SELECT 
                    u.Id AS FriendId,
                    u.Username
                FROM Friendships f
                JOIN Users u ON 
                    (u.Id = f.FriendId AND f.UserId = @UserId)
                    OR (u.Id = f.UserId AND f.FriendId = @UserId)
                WHERE 
                    (f.UserId = @UserId OR f.FriendId = @UserId)
                    AND f.Status = 'accepted'
            `);

        return res.status(200).json({ friends: result.recordset });
    } catch (err) {
        console.error('Get friends error:', err);
        return res.status(500).json({ message: 'Server error' });
    }
};